metals/src/main/scala/scala/meta/internal/metals/scalacli/ScalaCliServers.scala

```diff
@@ -0,0 +1,224 @@
+
+  def loaded(path: AbsolutePath): Boolean =
+    servers.exists(_.path.toNIO.startsWith(path.toNIO))
+
+  def loadedExactly(path: AbsolutePath): Boolean =
+    servers.exists(_.path == path)
+
+  def paths: Iterable[AbsolutePath] = servers.map(_.path)
+
+  def start(path: AbsolutePath): Future[Unit] = {
+    val customWorkspace =
+      if (path.isDirectory) None
+      else
+        Some {
+          val globalTmpDir =
+            scalaCliBuildDirectory.get() match {
+              case Some(workspace) => workspace
+              case None =>
+                val tmpFile =
+                  AbsolutePath(
+                    Files.createTempDirectory(s"metals-scala-cli")
+                  )
+                val Some(workspace) =
+                  scalaCliBuildDirectory.updateAndGet {
+                    case None => Some(tmpFile)
+                    case some => some
+                  }
+                workspace
+            }
+
+          // When path and workspace have different roots on Windows `scala-cli` throws an error,
+          // so we fallback to creating a tmp dir relative to `path`.
+          if (globalTmpDir.toNIO.getRoot() == path.toNIO.getRoot()) globalTmpDir
+          else {
+            val workspace = localTmpWorkspace(path)
+            if (!workspace.exists) Files.createDirectory(workspace.toNIO)
+            workspace
+          }
+        }
+    val scalaCli =
+      new ScalaCli(
+        compilers,
+        compilations,
+        statusBar,
+        buffers,
+        indexWorkspace,
+        diagnostics,
+        tables,
+        buildClient,
+        languageClient,
+        config,
+        cliCommand,
+        parseTreesAndPublishDiags,
+        path,
+        customWorkspace,
+      )
+
+    val prevServers = serversRef.getAndUpdate { servers =>
+      if (servers.exists(_.path == path)) servers
+      else {
+        if (servers.size == maxServers) servers.drop(1) :+ scalaCli
+        else servers :+ scalaCli
+      }
+    }
+
+    val (newServer, oldServer) =
+      prevServers
+        .find(_.path == path)
+        .map((_, None))
+        .getOrElse {
+          buildTargets.addData(scalaCli.buildTargetsData)
+          val oldServer =
+            Option.when(servers.size == 10)(prevServers.dequeue._1)
+          (scalaCli, oldServer)
+        }
+
+    for {
+      _ <- oldServer.map(_.stop()).getOrElse(Future.unit)
+      _ <- newServer.start()
+    } yield ()
+  }
+
+  def stop(): Future[Unit] = {
+    val servers = serversRef.getAndSet(Queue.empty)
+    Future.sequence(servers.map(_.stop()).toSeq).ignoreValue
+  }
+
+  def stop(path: AbsolutePath): Future[Unit] = {
+    val servers = serversRef.getAndUpdate(s => s.filterNot(_.path == path))
+    servers
+      .collectFirst {
+        case s if s.path == path =>
+          buildTargets.removeData(s.buildTargetsData)
+          s.stop()
+      }
+      .getOrElse(Future.successful(()))
+  }
+
+  def didFocus(path: AbsolutePath): Queue[ScalaCli] =
```

Comment:
```suggestion
  def didFocus(path: AbsolutePath): Unit =
```
I don't think it's used. Might be worth commenting that this is to make sure that older scala cli paths are discarded first.